### ISSUE TITLE
Check ordering of core exercises

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/csharp#611]
I just received some feedback from a student that he thought the `sum-of-multiples` exercise was way harder than the previous exercises. We should look into the current order and see if the difficulty is progressing as smoothly as possible.

**Comment by pdmoore** [comment]
Similarly, "Two Fer" comes after Bob and Raindrops. On the Java track "Two Fer" is the first exercise after Hello World. On C# track "Two Fer" is the 6th exercise and doesn't really add to what has already been covered.

**Comment by ErikSchierboom** [comment]
Thanks for the feedback! I'll come up with a new proposal soon.

**Comment by ErikSchierboom** [comment]
Some initial thoughts on this:

- `two-fer` should be much higher in the list, as it is actually quite easy.
- we should remove `sum-of-multiples` as a core exercise (see [this thread](https://github.com/exercism/exercism.io/issues/4110#issuecomment-410840861)).
- maybe `accumulate` isn't the best of core exercises, as people don't really learn much from it (except for the `yield` keyword).
- maybe `markdown` isn't a great core exercise, as it is primarily about refactoring and as such, does not teach any new skills.

**Comment by ErikSchierboom** [comment]
@jpreese @robkeim Your thoughts on this?

**Comment by robkeim** [comment]
Do we have data that we could use from before the release of Exercism v2 that would help us determine this order? Some values such as:
- The number of students that completed an exercise
- The number of iterations on average per exercise
- The order in which students completed the exercise

Might help give us more insight and the true "difficulty" of exercises as perceived by the students. With the release of v2 we can't trust the stats as much for the existing core exercises since we're forcing students into a specific order, but that wasn't the case in the first version.

**Comment by ErikSchierboom** [comment]
While working on the re-ordering, here are some of my thoughts on the current track's status:

- Some exercises unlock way more side exercises than others. This should be more spread out.
- We could probably add some more core exercises, to not unlock massive numbers of side exercises each time.
- We should probably add some of the more difficult exercises as core exercises.

**Comment by ErikSchierboom** [comment]
As a first step, this is how I currently envision that the exercises of difficulty `hard` should progress:

# Difficulty: hard
`diamond` (_core_): algorithms, strings
 -> `sgf-parsing`: parsing, strings
 -> `connect`: algorithms, strings 
 -> `go-counting`: algorithms, strings
`hangman` (_core_): reactive programming
 -> `react`: events
`zebra-puzzle` (_core_): logic, laziness
 -> `two-bucket`: logic
`forth` (_core_): stacks
 -> `alphametics`: algorithms
`zipper` (_core_): trees
 -> `pov`: trees

The core progression path is thus: 
`diamond` -> `hangman` -> `zebra-puzzle` -> `forth` -> `zipper`

Here is my rationale for this progression:

- I've tried to select those exercises as core exercises that are easier than the ones they unlock, and will probably have solutions that are interesting to mentor/receive feedback on. 
- The core exercises unlock similar side exercises, such that they are better related to the core. This will allow students to further hone their skills on a particular subject through the use of side exercises.
- Each core exercise unlocks a small set of side exercises, to not overwhelm the user with side exercises.
- The core exercises are quite diverse in their topics, with little overlap.

This does leave one hard exercise ungrouped: `variable-length-quantity`. I plan to have that unlocked as part of the medium difficulty exercises.

**Comment by ErikSchierboom** [comment]
@robkeim @pdmoore @jpreese The comment above contains my suggestion for an updated core exercise path. Any thoughts?

**Comment by ErikSchierboom** [comment]
@robkeim @jpreese Could you perhaps chime in with your thoughts on this thread: https://exercism-team.slack.com/messages/CASDGG84B/convo/CASDGG84B-1543225722.030400/

**Comment by robkeim** [comment]
@ErikSchierboom how do we get an access to that slack workspace?

**Comment by ErikSchierboom** [comment]
Hmmm, you're not an exercism mentor, are you? My bad. I'll just post the question here then:

> I'm currently doing some thinking on how to better structure the C# track (which exercises are core, which order, etc.). One of the problem I have is determining what are the key, basic types one has to learn in the first few exercises. Currently I have string, int, bool and datetime. I was wondering if some collection type should also be introduced as a basic type, as collections are very important. To me, the most obvious choice would be an array. However, there are not many array-based exercises at the moment. One array exercise that is fairly simple, is high-scores. However, the most simple implementation would require one to use LINQ, which in itself is actually a bit more advanced concept as it builds on extension methods. Do you think we could get away with introducing high-scores as a basic exercise and point people to LINQ very soon in the track (they wouldn't have to understand how it works, just be able to use it), or would that be too complicated?

There was one response so far:

> I've only done a handful of exercises, so take this with a grain of salt, but here are my initial thoughts:
> 1. Sum of Multiples (the fourth exercise in the track) ends up using LINQ pretty heavily. If we're worried about introducing LINQ, then we should move that one back further.
> 2. I'm frankly not all that afraid to introduce LINQ early on: it's possible that people won't grasp the nuances of extension methods and delegates, but the syntax and concepts are fairly straightforward: I would say more straightforward than new object construction and for loops. And it's used far more often in real C# code (in my experience) than for loops are.
> 3. Arrays are certainly very basic data structures, in the sense that they're closer to the metal, but in real life I think `List<>`s are more "basic" in the sense of being more commonly used and less likely to surprise you. If we're borrowing from a common set of exercises that assumes arrays are more fundamental than lists, then I'm not opposed to starting with arrays, but I wouldn't necessarily favor arrays as an earlier concept than lists, all else being equal.

To which I responsed:

> Thanks for the very detailed response!
> 
> 1. The sum-of-multiples exercise won't be a core exercise anymore, as there is a new policy that math-related exercises should not be core exercises.
> 2. Good to hear! I hope to hear from more people what their take on this is, but this is very hopeful to me.
> 3. I was really struggling with this, as I also initially considered List as the basic type. We can totally determine for ourselves if we prefer List over Array, so there is no pressure there.

**Comment by robkeim** [comment]
No I didn't join as a mentor when v2 was released.

Here's my $0.02 after having some time to think this over:

I agree that the notion of an enumerable should be introduced as a fundamental "data type" in C#. I don't think it makes a big difference whether we do that via an array or a list. That said, list might be slightly easier to understand since the notion of a list is something everyone is familiar with whereas the notion of an array is used only in mathematics.

I would be hesitant to introduce LINQ and particularly LINQ heavy exercises too early in the track. I think there are a few different "levels" of understanding LINQ:
1. I use LINQ but don't really understand the syntax
2. I use LINQ and understand concepts it uses (extension methods, lambda expressions, delegates)
3. I use LINQ and understand "how" it works (could implement it if asked to)

In my opinion, the jump to even get to the first level is pretty high particularly for someone that doesn't have SQL or functional programming experience. That said, if the "non-LINQ" equivalent solution isn't too complicated then leaving it early in the track shouldn't pose a problem as students can decide themselves whether they want to leverage LINQ or "write it by hand" so to speak.

**Comment by ErikSchierboom** [comment]
Thanks for the thoughtful feedback! Currently, the most basic list-based exercise is probably `high-scores`: https://github.com/exercism/problem-specifications/blob/master/exercises/high-scores/canonical-data.json. When this exercise is implemented, users could use the `Sum()` and `Take()` and `OrderBy()` LINQ methods to solve it. Do you think that would be too hard? If so, we might try to come up with a new, basic, list-based exercise.

**Comment by robkeim** [comment]
I think this exercise should be fine to solve without LINQ... although much more succinct using LINQ :)

Here's an untested proof of concept implementation I through together in 10 minutes only using List from the Collections namespace no other dependencies:
```
public class HighScores
{
    private List<int> _scores;

    public HighScores(List<int> scores)
    {
        _scores = new List<int>(scores);
    }

    public IEnumerable<int> Scores()
    {
        return _scores;
    }

    public int Latest()
    {
        return _scores[_scores.Count - 1];
    }

    public int PersonalBest()
    {
        var best = int.MinValue;

        foreach (var score in _scores)
        {
            if (score > best)
            {
                best = score;
            }
        }

        return best;
    }

    public List<int> PersonalTop()
    {
        var scores = new List<int>();
        scores.AddRange(_scores);
        scores.Sort();

        var result = new List<int>();

        var index = scores.Count - 1;

        while (result.Count < 3 && index >= 0)
        {
            result.Add(scores[index]);
            index--;
        }

        return result;
    }
}
```

**Comment by ErikSchierboom** [comment]
> In my opinion, the jump to even get to the first level is pretty high particularly for someone that doesn't have SQL or functional programming experience.

This was also my gut feeling.

Thanks for putting the example non-LINQ solution together. It has helped me see how difficult a non-LINQ implementation would be. Seeing this implementation, I think that this exercise is great to introduce LINQ, but it should probably not be a basic exercise (which means one of the five first exercises).

CC @F3PiX

**Comment by ErikSchierboom** [comment]
The first stage of the re-ordering has been merged in #740.

**Comment by ErikSchierboom** [comment]
With #744 merged, I'm gonna close this issue. There will be continuous re-evaluation of the track's ordering, but the major re-ordering is done.
